### ISSUE TITLE
get_perlin_map consumes all memory and brings down servers

### ISSUE THREAD

**Issue by bugreport0** [minetest/minetest#2665]
<b>Edit: Skip to summary: https://github.com/minetest/minetest/issues/2665#issuecomment-100410934</b>

I can't run a server properly due to this bug... the memory gets out of control and crashes minetest very often.

Summary:
get_perlin_map consumes insane amounts of memory more than any server can handle safely... much more than the measly 343KB that it actually contains.

Steps to Reproduce:
An easy way to see this in action is with a LUA mapgen, like watershed.
1. Download and install watershed:
https://forum.minetest.net/viewtopic.php?f=9&t=8609
2. Run a server
3. Connect to the server on another client, use fast, fly, and noclip... and explore the map as fast as you can.
4. Watch the maximum memory usage on the server process.  In testing, on my local machine I can go over 1GB.  On the server we are running the process OFTEN goes over 2GB and crashes the server.

Details of the problem:
I did some testing to try and narrow down the cause.  Here's what I found out:
* The problem affects various versions of minetest so appears to still be a problem.
* I removed all the code in watershed except for the get_perlin_map calls.  This shows that the memory issue is in the get_perlin_map functions.

For your reference, here are the actual perlin map calls from watershed:
	local nvals_terrain = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_fissure = minetest.get_perlin_map(np_fissure, chulens):get3dMap_flat(minposxyz)
	local nvals_temp = minetest.get_perlin_map(np_temp, chulens):get3dMap_flat(minposxyz)
	local nvals_humid = minetest.get_perlin_map(np_humid, chulens):get3dMap_flat(minposxyz)
	local nvals_seam = minetest.get_perlin_map(np_seam, chulens):get3dMap_flat(minposxyz)
	local nvals_strata = minetest.get_perlin_map(np_strata, chulens):get3dMap_flat(minposxyz)
	local nvals_mid = minetest.get_perlin_map(np_mid, chulens):get2dMap_flat(minposxz)
	local nvals_base = minetest.get_perlin_map(np_base, chulens):get2dMap_flat(minposxz)
	local nvals_xlscale = minetest.get_perlin_map(np_xlscale, chulens):get2dMap_flat(minposxz)
	local nvals_magma = minetest.get_perlin_map(np_magma, chulens):get2dMap_flat(minposxz)

Note: this is creating a big problem in running a stable server... it crashes so often.

**Comment by bugreport0** [comment]
Additional notes:
> If I understand correctly, a single perlin map should only contain 343KB of data.  7*7*7*16 * 64bit data = 343KB.  Although, I realize that the perlin map function may use up more memory when it is generating the data in the first place... the code is either not freeing up that memory after the map has been generated or there is some other issue....  For example, once the perlin map has been assigned to nvals_terrain variable, it should only be consuming 343KB of data and not several hundred MB.

**Comment by kwolekr** [comment]
Erm... I do not understand how you reached those figures for memory consumption at all.

Instead of pointing fingers at the engine, perhaps you should share with us how much ram your server has, and maybe the size of the pagefile?

**Comment by paramat** [comment]
Did you use the latest version of watershed? a couple of months ago it had an error that was consuming a lot of memory.

**Comment by bugreport0** [comment]
> Did you use the latest version of watershed?
I don't think this would matter.  I basically deleted ALL of watershed except for the 11 perlin map functions ... which is why I think it's related to those functions.

> Erm... I do not understand how you reached those figures for memory consumption at all.
Ran the game and watched RAM usage.  minetest was consuming over 1GB at times with the perlin map functions and much less RAM without the perlin map functions.

**Comment by bugreport0** [comment]
P.S. I'll see if I can narrow it down with a nice testcase... give me a little bit of time.

**Comment by bugreport0** [comment]
Here is a reduced version that shows the problem. Basically everything is removed but the perlin map calls.
```
-- watershed 0.5.0 by paramat
-- License: code WTFPL, textures CC BY-SA

-- 3D noise for terrain

local np_terrain = {
	offset = 0,
	scale = 1,
	spread = {x=384, y=192, z=384},
	seed = 593,
	octaves = 5,
	persist = 0.67
}

-- Mapchunk generation function

function watershed_chunkgen(x0, y0, z0, x1, y1, z1, area, data)
	-- perlinmap stuff
	local sidelen = x1 - x0 + 1 -- chunk sidelength
	local chulens = {x=sidelen, y=sidelen+2, z=sidelen} -- chunk dimensions, '+2' for overgeneration
	local minposxyz = {x=x0, y=y0-1, z=z0}
	local minposxz = {x=x0, y=z0}
	-- 3D and 2D perlinmaps
	local nvals_terrain = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_fissure = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_temp = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_humid = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_seam = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_strata = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
	local nvals_mid = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
	local nvals_base = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
	local nvals_xlscale = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
	local nvals_magma = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
end

minetest.register_on_generated(function(minp, maxp, seed)
	local t1 = os.clock()
	local x1 = maxp.x
	local y1 = maxp.y
	local z1 = maxp.z
	local x0 = minp.x
	local y0 = minp.y
	local z0 = minp.z
	--print ("[watershed] generate mapchunk minp ("..x0.." "..y0.." "..z0..")")
	watershed_chunkgen(x0, y0, z0, x1, y1, z1, area, data)
	local chugent = math.ceil((os.clock() - t1) * 1000)
	print ("[watershed] "..chugent.." ms")
end)


```

Here's what I did in testing:
1. Ran a server on a local machine using this file as a mod.
2. Connected to server and flew around the map very fast.  Basically, turn on fast, and hold down the "w" and "shift" keys.
3. Watched the memory usage on the server process.
4. Within about 30 seconds, if not sooner, the memory usually goes over 500MB... usually over 700MB and I have gotten over 1GB in the past.

**Comment by bugreport0** [comment]
Sidenote:
If you prefer, you can change the subject of this bug report to something less dramatic:
"get_perlin_map consuming an extreme amount of memory"

It has crashed the server many times in the past, but I realize you need more details to study the crash and that could take time.  Hopefully this would be enough for someone else to see if its a real bug or not.

**Comment by kwolekr** [comment]
Yes, the version of watershed does matter.  A rather important detail with the code you distilled the issue to was modified.
Again, I'll ask:
 - How on earth did you get the number "373 KB"?  That's outrageously incorrect and seems random.
 - How large is the swapfile on your computer, and how much memory do you have?  What operating system are you using?  If non-Windows, do you have a ulimit set for memory usage?

**Comment by paramat** [comment]
Yes try a more recent version of watershed that has chulens2d with z size = 1, apparently this change consumes much less memory.

**Comment by paramat** [comment]
Possibly related to this, when i mapgen dev and am flying at constant top speed generating terrain, my hard-drive seems to get very busy after a few minutes, constantly working instead of saving every few seconds, any idea if this is normal?

**Comment by paramat** [comment]
You could also try latest riverdev mod for the ultimate test, that has heavier use of noise.

**Comment by bugreport0** [comment]
1. Can you check my numbers?

Assume that each data point contains a 64bit floating point number.
Assume that you are working with one mapgen chunck, which is 7x7x7 blocks or 112x112x112 data points.
This means that each nvals variable should only be storing 112x112x112x64 bits ... or 11,239,424 bytes or 10MB?   ...well maybe a bit more since it's LUA
If so, in my testcase, there are 10 nvals, so that comes to around 100MB of data.
I also assume the mapgen system is singlethreaded (not the lua part, but the core) and only allowing one chunck at a time to be processed instead of several in parallel.  If this is not true, then that might be the cause of the high memory spikes.
Since I am seeing the server process consume 500-1GB, does this not suggest things are going wrong somehow?
2. Are you able to reproduce the same high memory usage using my testcase?
3. About your chulens suggestion...  Do you mean using NO overgeneration on the x,y, or z axis like this?
 local chulens = {x=x0-x1, y=x0-x1, z=x0-x1}
Basically you are saying limit yourself to the 5x5x5 blocks instead of using the 7x7x7 blocks the mapgen provides?
4. P.S. Even if not using overgeneration fixes the memory problem it would be nice if we could figure out what that causes the huge memory spike... as you know so many mapgens need that overgeneration feature. :)
5. Additional notes: I do not have enough details on how this bug affects the server.  Technically, the server process is not crashing, but being killed off by the OS or hosting company due to high memory usage.  Since it would be too problematic to test the bug on the server, I simply tested on my local machine to see what kind of memory usage minetest uses and was surpised to find extremely high memory spikes.  If someone can confirm that they are also able to observe the same problem using my testcase, that'd be great.
6. P.S. should I do some research into how local chulens = {x=x0-x1, y=x0-x1, z=x0-x1} affects memory?  I suppose I could also test to see if minetest is threading the chunck generation unless you know for sure it doesn't?

**Comment by bugreport0** [comment]
Well this doesn't seem to help much, but I might as well post it.
Memory usage tests ... create a local server and connect to it with a client, then hold down shift and "w" while flying fast.

1. baseline minetest using singlnode and no LUA mapgen:
345MB
Should it really be using that much for singlenode?  If we had a way to disable nyancats and other builtin mapgen additions, would this lower the memory usage?

Next, I tried something else, I completely turned off the LUA mapgen and tried this function:

'''
minetest.register_chatcommand("regen",{
	description = "Regenerate player's current mapchunk",
	privs = {privs = server, rollback},
	func = function(name, params)
		while name do
		local t1 = os.clock()
		local player = minetest.get_player_by_name(name)
		local pos = player:getpos()
		local plax = math.floor(pos.x + 0.5)
		local play = math.floor(pos.y + 0.5)
		local plaz = math.floor(pos.z + 0.5)
		local x0 = (80 * math.floor((plax + 32) / 80)) - 32
		local y0 = (80 * math.floor((play + 32) / 80)) - 32
		local z0 = (80 * math.floor((plaz + 32) / 80)) - 32
		local x1 = x0 + 79
		local y1 = y0 + 79
		local z1 = z0 + 79
		local sidelen = x1 - x0 + 1 -- chunk sidelength
		local chulens = {x=sidelen, y=sidelen, z=sidelen} -- chunk dimensions, '+2' for overgeneration
		local minposxyz = {x=x0, y=y0, z=z0}
		local minposxz = {x=x0, y=z0}
		-- 3D and 2D perlinmaps
		local nvals_terrain = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_fissure = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_temp = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_humid = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_seam = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_strata = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_mid = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_base = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_xlscale = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_magma = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local chugent = math.ceil((os.clock() - t1) * 1000)
		print ("[watershed] "..chugent.." ms")
		end
	end
})
'''

I'm no closer to an answer, but here's what I get:
1. Basline minetest = 40MB before I run these tests
2. If I use this code (overgen) and 10 nvals, memory spikes > 1.1GB
'''
	local sidelen = x1 - x0 + 1 -- chunk sidelength
	local chulens = {x=sidelen, y=sidelen+2, z=sidelen} -- chunk dimensions, '+2' for overgeneration
	local minposxyz = {x=x0, y=y0-1, z=z0}
'''
3. If I used the same code (overgen) and 1 nval, memory spikes at 650MB
4. If I use this code (no overgen) and 10 nvals, I get 763MB
'''
		local sidelen = x1 - x0 + 1 -- chunk sidelength
		local chulens = {x=sidelen, y=sidelen, z=sidelen} -- chunk dimensions, '+2' for overgeneration
		local minposxyz = {x=x0, y=y0, z=z0}
'''
5. If I use the same code (no overgen) and 1 nval, I get 433MB

<b>Conclusions: Even a minor overgen almost doubles the memory usage! ... just like you noted.
Even with the non-overgen tests, memory usage seems kinda high?</b>

Some problems:
I do not know what causes minetest to use over 2GB on the real server machine, when my tests can only reach around 1.2GB max.
Oddly, 10 nvals is not 10x the memory usage of 1 nval.

**Comment by paramat** [comment]
2. I'm going to try my own test with flying fast in a heavy mapgen.
3. No nothing to do with overgeneration. I mean use the latest version of watershed that has this code:

	-- perlinmap stuff
	local sidelen = x1 - x0 + 1 -- chunk sidelength
	local chulensxyz = {x=sidelen, y=sidelen+2, z=sidelen} -- chunk dimensions, '+2' for overgeneration
	local chulensxz = {x=sidelen, y=sidelen, z=1} -- here x = map x, y = map z
	local minposxyz = {x=x0, y=y0-1, z=z0}
	local minposxz = {x=x0, y=z0} -- here x = map x, y = map z

The version you have been using uses 2D perlinmaps with z dimension = sidelen instead of z = 1, this causes 240 times more memory use for 2D perlinmaps. However that may not be the entire cause of your problems. Please try to reproduce with latest watershed.
6. No don't bother as it's nothing to do with overgeneration.

**Comment by bugreport0** [comment]
After more testing, <b>I think I may have something worth considering....</b>

1. First, let me provide you with the test code.
```
-- watershed 0.5.0 by paramat
-- License: code WTFPL, textures CC BY-SA

-- 3D noise for terrain

local np_terrain = {
	offset = 0,
	scale = 1,
	spread = {x=384, y=192, z=384},
	seed = 593,
	octaves = 5,
	persist = 0.67
}

minetest.register_chatcommand("regen",{
	description = "Regenerate player's current mapchunk",
	privs = {privs = server, rollback},
	func = function(name, params)
		while name do
		local t1 = os.clock()

		local player = minetest.get_player_by_name(name)
		local pos = player:getpos()
		local plax = math.floor(pos.x + 0.5)
		local play = math.floor(pos.y + 0.5)
		local plaz = math.floor(pos.z + 0.5)
		local x0 = (80 * math.floor((plax + 32) / 80)) - 32
		local y0 = (80 * math.floor((play + 32) / 80)) - 32
		local z0 = (80 * math.floor((plaz + 32) / 80)) - 32
		local x1 = x0 + 79
		local y1 = y0 + 79
		local z1 = z0 + 79

		local sidelen = x1 - x0 + 1 -- chunk sidelength
		local chulens = {x=sidelen, y=sidelen, z=sidelen} -- chunk dimensions, '+2' for overgeneration
		local minposxyz = {x=x0, y=y0, z=z0}
		local minposxz = {x=x0, y=z0}
		-- 3D and 2D perlinmaps
		local nvals_terrain = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_fissure = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_temp = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_humid = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_seam = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_strata = minetest.get_perlin_map(np_terrain, chulens):get3dMap_flat(minposxyz)
		local nvals_mid = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_base = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_xlscale = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local nvals_magma = minetest.get_perlin_map(np_terrain, chulens):get2dMap_flat(minposxz)
		local chugent = math.ceil((os.clock() - t1) * 1000)
		print ("[watershed] "..chugent.." ms")
		end
	end
})

```

2. Now, run a local server, connect to it and run the "/regen" command.
3. Results: memory spikes rather high.

HOWEVER, here's the interesting part.

4. Comment out the "while" loop (comment out two lines) so it does not run in an infinite loop, but instead just runs one time.
5. Run the /regen command 1 time and look at the server process.  Memory usage goes up... <b>but then goes down after a few seconds.</b>
6. The key is that part in bold: It takes several seconds to free up the memory!
7. If you type "/regen" fast enough, memory will keep going up as it does not free up the old memory fast enough.  This is exactly what would happen in a real game when a player is exploring and generating many chuncks of the map one after the other.
8. <b>Result: Is it possible that the memory is not being freed fast enough from the get_perlin_map internals?... causing minetest memory to skyrocket during heavy map exploration... aka generating many chuncks one after another.</b>

Note: I still do not know why overgen uses 2x more memory, but maybe this could be a reason... or that may even be a totally separate issue.

**Comment by paramat** [comment]
> Result: Is it possible that the memory is not being freed fast enough from the get_perlin_map internals?... causing minetest memory to skyrocket during heavy map exploration... aka generating many chuncks one after another.

This might explain why my drive gets busy during a long session of flying and generating.

**Comment by bugreport0** [comment]
If the cause is related to get_perlin_map and it taking a long time to free the memory... there's not much else I can do do test this as someone who knows the internals of minetest would probably have to check on this. :(

**Comment by bugreport0** [comment]
> This might explain why my drive gets busy during a long session of flying and generating.
Sidenote:
The minetest client (not server) continues to consume an infinite amount of memory and then eventually crashes every single time I explore the map with fast and fly ... if that's what you mean, yeah I'm quite familiar with that problem... and it's causes me a great deal of problems.

However, I just want to be clear that the problem I mentioned in my previous post applies to the "server process" run from the commandline and not the client program with a GUI.

**Comment by paramat** [comment]
Stuff from IRC:
18:04 	hmmmm 	oh jeez
18:04 		i know, it's the lua table that's consuming the majority of the memory
18:04 		but it shouldn't be taking up gigabytes... no way
18:05 		how on earth do you measure consumption of lua?

18:07 	hmmmm 	it's funny, i profiled this a while ago.  writing the noise values to a lua table takes up about 10x as much CPU as generating them does

**Comment by paramat** [comment]
> Possibly related to this, when i mapgen dev and am flying at constant top speed generating terrain, my hard-drive seems to get very busy after a few minutes, constantly working instead of saving every few seconds, any idea if this is normal?

> This might explain why my drive gets busy during a long session of flying and generating.

Oh man ignore these comments, these are for core mapgen, i confused the two =/

**Comment by bugreport0** [comment]
Since some others are looking at this. I just wanted to summarize.  So far, I am seeing two big memory problems:

1. The problem paramat mentioned, where using an overgen of even just 2 causes memory usage to double vs using no overgen (5x5x5 blocks).
2. The problem where it literally takes several seconds to free up memory after calling get_perlin_map ... thus resulting in memory going up very high if you call many get_perlin_maps in a row like happens when generating terrain.

**Comment by bugreport0** [comment]
Hate to keep doing this, but there's actually a third concern I have:
3. If you have no mapgen (just singlenode), minetest server can consume 345MB for just one player who is flying fast!  Does this seem like an awefully high memory usage when you are just generating singlenode and nyan cats?

**Comment by paramat** [comment]
> The problem paramat mentioned, where using an overgen of even just 2 causes memory usage to double vs using no overgen (5x5x5 blocks).

I didn't mention this.

My rfirst results:
Using latest watershed in 0.4.12 stable and latest dev, flying fast and constantly generating, my memory use due to Minetest rises to 1.5GB often, but never more than that. Perhaps with the heavy use of 3D noise and the effect of lua tables this is actually reasonable? Luckily i have 8GB of RAM so this is not a problem.

**Comment by bugreport0** [comment]
Now, I'm confused, are you speaking about the client or the server?

<b>I am only referring to the server process in my notes.</b>  I would be curious if the server process you have gets that high memory usage?  If you do, that would be very interesting information ... especially under what situations you get that.

The important part is the server process especially because: servers have a limited number of RAM that is strictly enforced by most hosting companies, the server costs money, and many people rely on them being stable to play the game.
So it is REALLY important to use as little RAM as possible on the server... which is what I brought out my investigation into this problem.

> I didn't mention this.

I thought that's what you meant in your earlier posts.  Anyways, the tests show that overgen doubles the memory usage for whatever reason.  What I mean by this is that <b>if I get_perlin map on 80x80x80 blocks this will use half the memory than if I get_perlin_map on 80x82x80 blocks!</b>

**Comment by bugreport0** [comment]
I have question?  What would be the best way to summarize the details of my testing so that someone has all the info they need to investigate this further?  I noticed that there was some confusion and there is tons of details to wade through in this bug report.

Should I contact someone via private message and send them the details? If so who?
Should I edit the first post with the summary?
Or should I just post a reply back here, increasing the chance that the important details might get overlooked in all this talk?

**Comment by bugreport0** [comment]
I know I've said all of this in previous posts, but I suspect nobody wants to wade through all that information trying to find the relevant parts.  I figure the best way is to summarize the 3 most relevant parts....

Summary:

Various parts of the the minetest mapgen seem to be consuming more memory than they should be... perhaps too much for a normal VPS server to handle.  This issue is about the server process and not about the client/GUI program.

Why this matters so much too me:

To me the lua mapgens are the most fun to play.  However, there seems to be issues causing the memory to spike quite high for a short time and exceed the RAM limits most hosting companies places on servers.  This, of course, causes the server process to be shut down by the hosting company or OS.

Details:
I do not know the exact cause, but I have been able to narrow down three very specific cases that may help.  I will provide a summary, and then details on how to reproduce the bug on your system.

CASE 1:
Theory: Minetest takes too long to free memory from the get_perlin_map function, resulting in skyrocketing RAM usage if you call many get_perlin_map functions within a short time.
CASE 2:
A normal mapgen generates 80x80x80 pieces of data.  However, minetest also includes an "overgen" ability to generate 112x112x112 blocks so it can merge smoothly with nearby chuncks of the map.  Using get_perlin_map on 80x82x80 blocks consumes 2x the memory it does with 80x80x80 blocks.
CASE 3:
If you run minetest with singlenode and no lua mapgen, then the process consumes up to 345MB for just one player.  This seems like it may be an issue if the base without any mapgen spikes this high.

CASE 1 details:
Provided here: https://github.com/minetest/minetest/issues/2665#issuecomment-100385547
It seems as though minetest is not freeing up memory fast enough, but that is only a guess.
CASE 2 details:
Provided here: https://github.com/minetest/minetest/issues/2665#issuecomment-100377528
In summary, if I run get_perlin_map on 80x80x80 blocks it will consume half the memory of running get_perlin_map on 80x82x80 blocks ... just a few blocks more.
CASE 3 details:
1. Run a server with singlenode in the console
2. Connect to the server with a client program.
3. Fly around the map very fast.
4. Watch the peak memory usage on the minetest server process.  It went as high as 345MB just for one player.  That's without any mapgen whatsoever.  This may not be a bug, but might be worth checking.

**Comment by paramat** [comment]
> I am only referring to the server process in my notes.

Ah okay, my tests are using singleplayer, so client and server on one machine, so my system monitor program is showing the combined usage.

**Comment by Zeno-** [comment]
Game: the minimal minetest game
Mods: Just the test case above (with the infinite loop)
Blank world

It seems to me that the delay is just normal Lua garbage collector behaviour (??)

Tested with 
```
valgrind --tool=massif \
	--detailed-freq=1000000 --depth=7 --alloc-fn=g_malloc --alloc-fn=g_realloc \
	--alloc-fn=g_try_malloc --alloc-fn=g_malloc0 --alloc-fn=g_mem_chunk_alloc \
	\
	/home/RebeccaSnoot/minetest/bin/minetest --server \
	--config ~/.minetest/disktest.conf --world ~/.minetest/worlds/disk-minimal
```

![21255](https://cloud.githubusercontent.com/assets/5899883/7548643/7717f0bc-f655-11e4-8b4b-1068fdd3cd19.jpg)


Peak snapshot
```
time=240428969327
mem_heap_B=985028452
mem_heap_extra_B=661364
mem_stacks_B=0
heap_tree=peak
n5: 985028452 (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
 n2: 559744672 0xA602C9: l_alloc (lauxlib.c:636)
  n2: 559744056 0xA686D3: luaM_realloc_ (lmem.c:79)
   n2: 559159136 0xA7288E: setarrayvector (ltable.c:265)
    n2: 559153920 0xA72A99: resize (ltable.c:303)
     n1: 559153920 0xA72D9E: rehash (ltable.c:348)
      n2: 559153920 0xA72FBF: newkey (ltable.c:405)
       n0: 559153776 0xA73451: luaH_setnum (ltable.c:515)
       n0: 144 in 2 places, all below massif's threshold (01.00%)
     n0: 0 in 1 place, below massif's threshold (01.00%)
    n0: 5216 in 1 place, below massif's threshold (01.00%)
   n0: 584920 in 24 places, all below massif's threshold (01.00%)
  n0: 616 in 1 place, below massif's threshold (01.00%)
 n1: 194813952 0x9D686B: Noise::allocBuffers() (noise.cpp:507)
  n2: 194813952 0x9D66C0: Noise::Noise(NoiseParams*, int, int, int, int) (noise.cpp:475)
   n1: 194560000 0x78FA9C: LuaPerlinNoiseMap::LuaPerlinNoiseMap(NoiseParams*, int, irr::core::vector3d<short>) (l_noise.cpp:152)
    n1: 194560000 0x76F5E7: ModApiEnvMod::l_get_perlin_map(lua_State*) (l_env.cpp:681)
     n1: 194560000 0xA62F0E: luaD_precall (ldo.c:319)
      n1: 194560000 0xA77CAC: luaV_execute (lvm.c:587)
       n0: 194560000 0xA631B3: luaD_call (ldo.c:377)
   n0: 253952 in 8 places, all below massif's threshold (01.00%)
 n1: 194813952 0x9D689E: Noise::allocBuffers() (noise.cpp:508)
  n2: 194813952 0x9D66C0: Noise::Noise(NoiseParams*, int, int, int, int) (noise.cpp:475)
   n1: 194560000 0x78FA9C: LuaPerlinNoiseMap::LuaPerlinNoiseMap(NoiseParams*, int, irr::core::vector3d<short>) (l_noise.cpp:152)
    n1: 194560000 0x76F5E7: ModApiEnvMod::l_get_perlin_map(lua_State*) (l_env.cpp:681)
     n1: 194560000 0xA62F0E: luaD_precall (ldo.c:319)
      n1: 194560000 0xA77CAC: luaV_execute (lvm.c:587)
       n0: 194560000 0xA631B3: luaD_call (ldo.c:377)
   n0: 253952 in 8 places, all below massif's threshold (01.00%)
 n1: 32358400 0x96F914: MapBlock::reallocate() (mapblock.h:125)
  n1: 32358400 0x9736F6: MapBlock::MapBlock(Map*, irr::core::vector3d<short>, IGameDef*, bool) (mapblock.cpp:64)
   n2: 32358400 0x9A1E6C: MapSector::createBlankBlockNoInsert(short) (mapsector.cpp:92)
    n1: 31817728 0x9A1EBD: MapSector::createBlankBlock(short) (mapsector.cpp:99)
     n1: 31817728 0x96AD3C: ServerMap::createBlock(irr::core::vector3d<short>) (map.cpp:2703)
      n1: 31817728 0x96A165: ServerMap::initBlockMake(BlockMakeData*, irr::core::vector3d<short>) (map.cpp:2235)
       n0: 31817728 0x897E50: EmergeThread::getBlockOrStartGen(irr::core::vector3d<short>, MapBlock**, BlockMakeData*, bool) (emerge.cpp:413)
    n0: 540672 in 1 place, below massif's threshold (01.00%)
 n0: 3297476 in 295 places, all below massif's threshold (01.00%)
```

**Comment by Zeno-** [comment]
A longer test. The peaks are 20 iterations of the test code provided by @bugreport0 above. No leak per se, but I do have to admit that's quite a lot of RAM to be using.

![22188](https://cloud.githubusercontent.com/assets/5899883/7548866/60a081c2-f660-11e4-9138-d0dc94fdca7d.jpg)

```
#-----------
snapshot=22
#-----------
time=142901719139
mem_heap_B=931463756
mem_heap_extra_B=762508
mem_stacks_B=0
heap_tree=peak
n5: 931463756 (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
 n2: 525927101 0xA602C9: l_alloc (lauxlib.c:636)
  n2: 525926485 0xA686D3: luaM_realloc_ (lmem.c:79)
   n2: 525342560 0xA7288E: setarrayvector (ltable.c:265)
    n2: 525337344 0xA72A99: resize (ltable.c:303)
     n1: 525337344 0xA72D9E: rehash (ltable.c:348)
      n2: 525337344 0xA72FBF: newkey (ltable.c:405)
       n1: 525337200 0xA73451: luaH_setnum (ltable.c:515)
        n2: 525337200 0xA5DDE2: lua_rawseti (lapi.c:690)
         n1: 520093696 0x7901C2: LuaPerlinNoiseMap::l_get3dMap_flat(lua_State*) (l_noise.cpp:259)
          n1: 520093696 0xA62F0E: luaD_precall (ldo.c:319)
           n1: 520093696 0xA77CAC: luaV_execute (lvm.c:587)
            n1: 520093696 0xA631B3: luaD_call (ldo.c:377)
             n1: 520093696 0xA5E1B1: f_call (lapi.c:800)
              n1: 520093696 0xA62204: luaD_rawrunprotected (ldo.c:116)
               n1: 520093696 0xA635AC: luaD_pcall (ldo.c:463)
                n1: 520093696 0xA5E260: lua_pcall (lapi.c:821)
                 n1: 520093696 0x753C2B: script_run_callbacks(lua_State*, int, RunCallbacksMode) (c_internal.cpp:109)
                  n1: 520093696 0x7668E2: ScriptApiServer::on_chat_message(std::string const&, std::string const&) (s_server.cpp:139)
                   n1: 520093696 0x7384C4: Server::handleCommand_ChatMessage(NetworkPacket*) (serverpackethandler.cpp:1094)
                    n0: 520093696 0xA108D2: Server::handleCommand(NetworkPacket*) (server.cpp:1152)
         n0: 5243504 in 4 places, all below massif's threshold (01.00%)
       n0: 144 in 2 places, all below massif's threshold (01.00%)
     n0: 0 in 1 place, below massif's threshold (01.00%)
    n0: 5216 in 1 place, below massif's threshold (01.00%)
   n0: 583925 in 24 places, all below massif's threshold (01.00%)
  n0: 616 in 1 place, below massif's threshold (01.00%)
 n1: 190717952 0x9D686B: Noise::allocBuffers() (noise.cpp:507)
  n2: 190717952 0x9D66C0: Noise::Noise(NoiseParams*, int, int, int, int) (noise.cpp:475)
   n1: 190464000 0x78FA9C: LuaPerlinNoiseMap::LuaPerlinNoiseMap(NoiseParams*, int, irr::core::vector3d<short>) (l_noise.cpp:152)
    n1: 190464000 0x76F5E7: ModApiEnvMod::l_get_perlin_map(lua_State*) (l_env.cpp:681)
     n1: 190464000 0xA62F0E: luaD_precall (ldo.c:319)
      n1: 190464000 0xA77CAC: luaV_execute (lvm.c:587)
       n1: 190464000 0xA631B3: luaD_call (ldo.c:377)
        n1: 190464000 0xA5E1B1: f_call (lapi.c:800)
         n1: 190464000 0xA62204: luaD_rawrunprotected (ldo.c:116)
          n1: 190464000 0xA635AC: luaD_pcall (ldo.c:463)
           n1: 190464000 0xA5E260: lua_pcall (lapi.c:821)
            n1: 190464000 0x753C2B: script_run_callbacks(lua_State*, int, RunCallbacksMode) (c_internal.cpp:109)
             n1: 190464000 0x7668E2: ScriptApiServer::on_chat_message(std::string const&, std::string const&) (s_server.cpp:139)
              n1: 190464000 0x7384C4: Server::handleCommand_ChatMessage(NetworkPacket*) (serverpackethandler.cpp:1094)
               n1: 190464000 0xA108D2: Server::handleCommand(NetworkPacket*) (server.cpp:1152)
                n1: 190464000 0xA0421A: Server::ProcessData(NetworkPacket*) (server.cpp:1229)
                 n1: 190464000 0xA032FF: Server::Receive() (server.cpp:1027)
                  n1: 190464000 0x9FD8D7: ServerThread::Thread() (server.cpp:109)
                   n1: 190464000 0x6F9AEF: JThread::TheThread(void*) (jthread.cpp:157)
                    n0: 190464000 0x76D3528: start_thread (in /usr/lib64/libpthread-2.20.so)
   n0: 253952 in 8 places, all below massif's threshold (01.00%)
 n1: 190717952 0x9D689E: Noise::allocBuffers() (noise.cpp:508)
  n2: 190717952 0x9D66C0: Noise::Noise(NoiseParams*, int, int, int, int) (noise.cpp:475)
   n1: 190464000 0x78FA9C: LuaPerlinNoiseMap::LuaPerlinNoiseMap(NoiseParams*, int, irr::core::vector3d<short>) (l_noise.cpp:152)
    n1: 190464000 0x76F5E7: ModApiEnvMod::l_get_perlin_map(lua_State*) (l_env.cpp:681)
     n1: 190464000 0xA62F0E: luaD_precall (ldo.c:319)
      n1: 190464000 0xA77CAC: luaV_execute (lvm.c:587)
       n1: 190464000 0xA631B3: luaD_call (ldo.c:377)
        n1: 190464000 0xA5E1B1: f_call (lapi.c:800)
         n1: 190464000 0xA62204: luaD_rawrunprotected (ldo.c:116)
          n1: 190464000 0xA635AC: luaD_pcall (ldo.c:463)
           n1: 190464000 0xA5E260: lua_pcall (lapi.c:821)
            n1: 190464000 0x753C2B: script_run_callbacks(lua_State*, int, RunCallbacksMode) (c_internal.cpp:109)
             n1: 190464000 0x7668E2: ScriptApiServer::on_chat_message(std::string const&, std::string const&) (s_server.cpp:139)
              n1: 190464000 0x7384C4: Server::handleCommand_ChatMessage(NetworkPacket*) (serverpackethandler.cpp:1094)
               n1: 190464000 0xA108D2: Server::handleCommand(NetworkPacket*) (server.cpp:1152)
                n1: 190464000 0xA0421A: Server::ProcessData(NetworkPacket*) (server.cpp:1229)
                 n1: 190464000 0xA032FF: Server::Receive() (server.cpp:1027)
                  n1: 190464000 0x9FD8D7: ServerThread::Thread() (server.cpp:109)
                   n1: 190464000 0x6F9AEF: JThread::TheThread(void*) (jthread.cpp:157)
                    n0: 190464000 0x76D3528: start_thread (in /usr/lib64/libpthread-2.20.so)
   n0: 253952 in 8 places, all below massif's threshold (01.00%)
 n1: 13959168 0x96F914: MapBlock::reallocate() (mapblock.h:125)
  n1: 13959168 0x9736F6: MapBlock::MapBlock(Map*, irr::core::vector3d<short>, IGameDef*, bool) (mapblock.cpp:64)
   n1: 13959168 0x9A1E6C: MapSector::createBlankBlockNoInsert(short) (mapsector.cpp:92)
    n0: 13959168 in 2 places, all below massif's threshold (01.00%)
 n0: 10141583 in 293 places, all below massif's threshold (01.00%)
```

**Comment by bugreport0** [comment]
> It seems to me that the delay is just normal Lua garbage collector behaviour (??)

This is my thoughts exactly!  The pattern looks exactly like something you'd expect from garbage collection.  Basically, it seems like it is taking several seconds for garbage collection to remove the data from memory... even though the function has long expired.
It could either be the lua garbage collection or garbage collection in the minetest c++.

Give me some time to do some proper testing on that.  I really think that might be getting close to the source of the issue....

**Comment by bugreport0** [comment]
To kwolekr:
Early on you ask me many questions which I ignored. Sorry.  If you still would like me to address any of them just let me know.  I do not have root access to a Linux machine. I realize that would be ideal if I tested on a Linux server; it would give us both more confidence in what is going on.  However, all the testing data I have provided you was on a Windows machine.

**Comment by bugreport0** [comment]
GREAT NEWS!
CASE 1 and CASE 2 solved and not caused by core minetest! CASE 3 still unsolved.  May or may not be in core minetest.

CASE 1: memory taking to long to be freed.
This issue was indeed the lua garbage collector.  It was taking too long to free the nvals, causing a dramatic increase in memory usage.
Details:
Using the testcase from: https://github.com/minetest/minetest/issues/2665#issuecomment-100385547
I get the following results:
753,064 KB max memory
110-125ms mapgen time
However, if I add a "collectgarbage()" call to the very end of the function, I get:
111,184 KB max memory usage!
125ms mapgen time
You can even reduce the memory usage even more by inserting 11 "collectgarbage()" calls before and after EVERY nval/get_perlin_map.  The results of doing this are:
65,164 KB max memory usage!
141ms mapgen time

CASE 2: use of overgen uses 2x more memory
It seems like this *might* also be related to the lua garbage collection.
Details:
I used the method where I ran "collectgarbage()" before and after every get_perlin_map.  The results were:
65,164 KB without overgen
77,924 KB with overgen!
Although, the mapgen time got worse:
141ms without overgen
172-200ms with overgen!

CASE 3: singlenode (no mapgen) still uses up too much memory, maybe?
Well, the big memory problems are now solved, I hope, but I wonder if maybe we could still investigate CASE 3?

**Comment by bugreport0** [comment]
About CASE 3:

I did a test, where I removed the on_generated from the minetest game, so that it doens't even generate nyan cats and other stuff.  I think this removes all LUA mapgen from the game??

Even after doing this (supposedly I removed all LUA mapgen stuff), the memory in minetest can still go over 300MB in the right cases.

1. Run a server with singlenode in the console
2. Connect to the server with a client program.
3. Fly around the map very fast.
4. Watch the peak memory usage on the minetest server process. It went as high as 345MB just for one player.
5. The pattern appears to be the same as what happened with CASE 1, where the game is taking several seconds to free up the memory and if you generate enough chuncks fast enough the memory increases a lot.

I have no idea if this a bug or not ... maybe it's something similar to the lua garbage collection but in core? ... or maybe the database needs that much memory.... I don't know.  (If it's the database, then I guess it's not a bug, as maybe sqlite needs that much to properly cache the data because it can't write to disk fast enough?)

**Comment by bugreport0** [comment]
Sidenote:
I wonder if we might be able to reduce memory a tad more?  ... which is always good for servers.

<b>Could someone who works with the C++ core, take a look at the get_perlin_map and noise functions to make sure they are freeing memory manually right away instead of waiting on garbage collection?</b>  This is NOT a high priority task, but just something that might be worth investigating sometime.  I only suggest this because maybe it might mean we could use even less memory. :)

**Comment by paramat** [comment]
The earlier version of watershed you have been testing has 2D perlinmaps with z size of 82 instead of 1, which apparently multiplies memory use by 240. Also that version creates new noise objects for every mapchunk, instead of creating them once during the first mapchunk only, this also uses more memory.
Both these issues are fixed in the latest version, which is why i earlier asked which version you are using. However that is not the entire cause of the high memory use, as i have tested the latest watershed and still get a high memory usage. If you follow IRC logs for minetest-dev channel you can see hmmmm is working on improving the situation. This issue is not a bug but my monster mapgens with heavy use of noise in lua are showing up a problematicly high use of memory.
Note that watershed and riverdev both have heavier use of noise than any core mapgen, and doing so in lua makes the situation even worse, i notice my computers work much harder running these mods than core mapgen.

**Comment by bugreport0** [comment]
I'm not sure how I'm supposed to answer....  I think we might be on the same page on most things, but I'm not quite sure....

Just to be clear about a few things:

1. The reason I never addressed any specific version of watershed or other mapgens is that I did understand why you were asking about it, so I didn't think it was relevant.... sorry.... and I also am concerned about this issue from the perspective of writing custom LUA mapgens and not so much about watershed or any specific mapgen.

2. I now know that the major memory issues happen because of the LUA garbage collector.  CASES 1 and 2 are entirely caused by the LUA garbage collector and/or how you use the code.

3. 
> Also that version creates new noise objects for every mapchunk, instead of creating them once during the first mapchunk only, this also uses more memory.

Thanks. I think I finally understand why you mentioned this about the newer watershed.  If it does like you say, then that means it completely avoids the LUA garbage collector problem, so it probably doesn't have the major memory issues in that version.  Interesting news; thanks again.
4. <b>Sadly, CASE 3 remains unsolved (which can cause minetest to skyrocket above 300MB).</b>  As far as I can tell this is not related to LUA since it happens when you remove all lua mapgen stuff.

Unfortunately, a core dev (on the c++ side) will need to investigate CASE 3. :(  Any takers? :)
https://github.com/minetest/minetest/issues/2665#issuecomment-100555250

**Comment by paramat** [comment]
>  it completely avoids the LUA garbage collector problem, so it probably doesn't have the major memory issues in that version

Garbage collection is still used, and latest watershed still has a very high memory use, i tested it. Hmmmm/kwolekr is away for a week but has started work on this issue and has a solution that reduces memory use.

**Comment by paramat** [comment]
See 3ffb5f5761a83773037869d6f6179353c46a650a and 4c9a8a91c4988b3567a38af622a3eb0d0ec19f6b and https://forum.minetest.net/viewtopic.php?f=47&t=12178
I will edit my 2 monster mapgens to use and test this soonish. Perhaps this issue can be closed now?
